mantidproject/mantid
# Loading and saving of single value ws

**Description of work.**

This PR enables saving and loading of single value workspaces (`WorkspaceSingleValue`) to address the found issue where this type of workspaces would not be saved with the project, causing issues in recuperating reduction workflow started in a separate session. During loading, the proposed differentiation between a `Workspace2D` that has one bin and one spectrum from the special type of `WorkspaceSingleValue` is that if a workspace to be loaded contains no metadata (except for the default group included, called 'goniometer') AND has no instrument attached (instrument being an empty string), then it is effectively a single value workspace.

This PR contains also contributions from @MathieuTillet . 



**To test:**

1. Run the following script:
```
from mantid.simpleapi import *
import numpy as np

sv_ws = 'singlevalue_ws'
CreateSingleValuedWorkspace(15, OutputWorkspace=sv_ws)

sv_like_ws = '2dws_w_singlevalue'
CreateWorkspace(DataX=np.array([0]), DataY=np.array([5]), NSpec=1, OutputWorkspace=sv_like_ws)

print(type(mtd[sv_ws]))      # this should be WorkspaceSingleValue
print(type(mtd[sv_like_ws])) # this should be Workspace2D
path1=f'{sv_ws}_nxps.nxs'    # nexus processed
SaveNexusProcessed(InputWorkspace=sv_ws, Filename=path1)
path2=f'{sv_ws}_nxs.nxs' # nexus
SaveNexus(InputWorkspace=sv_ws, Filename=path2)
path3=f'{sv_like_ws}_nxps.nxs' # nexus processed
SaveNexusProcessed(InputWorkspace=sv_like_ws, Filename=path3)
path4=f'{sv_like_ws}_nxs.nxs' # nexus
SaveNexus(InputWorkspace=sv_like_ws, Filename=path4)

LoadNexus(Filename=path1, OutputWorkspace=f'{sv_ws}_nxps_loaded')
LoadNexus(Filename=path2, OutputWorkspace=f'{sv_ws}_nxs_loaded')
LoadNexus(Filename=path3, OutputWorkspace=f'{sv_like_ws}_nxps_loaded')
LoadNexus(Filename=path4, OutputWorkspace=f'{sv_like_ws}_nxs_loaded')

print(type(mtd[f'{sv_ws}_nxps_loaded']))      # should be still WorkspaceSingleValue
print(type(mtd[f'{sv_ws}_nxs_loaded']))       # should be still WorkspaceSingleValue
print(type(mtd[f'{sv_like_ws}_nxps_loaded'])) # should be still Workspace2D
print(type(mtd[f'{sv_like_ws}_nxs_loaded']))  # should be still Workspace2D
```

2. There should be no error mentioned in the issue #34021 , and the types should agree with the comments next to the line.
3. Save the project, and clear all workspaces in the ADS
4. Load the saved project: all workspaces previously present in the ADS should be properly loaded and the single valued workspace should not be missing and should be of the proper type.


Fixes #34021. 



Release notes added.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
